src/dp_maglev.c

```diff
@@ -0,0 +1,210 @@
+	}
+	return permutation;
+}
+
+static int dp_maglev_populate(struct lb_value *lbval, int *permutation)
+{
+	int *next = rte_zmalloc("maglev_hash", lbval->back_end_cnt * sizeof(int), RTE_CACHE_LINE_SIZE);
+	int i, j, pos, num = 0;
+
+	if (!next)
+		return DP_ERROR;
+
+	for (i = 0; i < DP_LB_MAGLEV_LOOKUP_SIZE; i++)
+		lbval->maglev_hash[i] = DP_MAGLEV_POS_FREE;
+
+	while (true) {
+		for (i = 0; i < lbval->back_end_cnt; i++) {
+			for (j = next[i]; j < DP_LB_MAGLEV_LOOKUP_SIZE; j++) {
+				pos = permutation[i*DP_LB_MAGLEV_LOOKUP_SIZE + j];
+				next[i]++;
+				if (lbval->maglev_hash[pos] == DP_MAGLEV_POS_FREE) {
+					lbval->maglev_hash[pos] = (int16_t)i;
+					num++;
+					if (num == DP_LB_MAGLEV_LOOKUP_SIZE)
+						goto out;
+					break;
+				}
+			}
+		}
+	}
+
+out:
+	rte_free(next);
+	return DP_OK;
+}
+
+static int dp_maglev_calc_hash_lookup_table(struct lb_value *lbval)
+{
+	int *permutation = dp_maglev_permutation(lbval);
+	int ret = DP_OK;
+
+	if (!permutation)
+		return DP_ERROR;
+
+	if (DP_FAILED(dp_maglev_populate(lbval, permutation))) {
+		DPS_LOG_WARNING("Loadbalancer Maglev population failed");
+		ret = DP_ERROR;
+	}
+
+	rte_free(permutation);
+	return ret;
+}
+
+static bool dp_is_ip_greater(const uint8_t *ip1, const uint8_t *ip2)
+{
+	int i;
+
+	for (i = 0; i < DP_IPV6_ADDR_SIZE; i++) {
+		if (ip1[i] < ip2[i])
+			return false;
+		if (ip1[i] > ip2[i])
+			return true;
+	}
+	return false;
+}
+
+static void dp_shift_back_end_ips(uint8_t ips[][DP_IPV6_ADDR_SIZE], int start_pos, int end_pos, bool shift_up)
+{
+	int i;
+
+	if (shift_up)
+		for (i = end_pos; i > start_pos; i--)
+			rte_memcpy(&ips[i], &ips[i - 1], DP_IPV6_ADDR_SIZE);
+	else
+		for (i = start_pos; i < end_pos; i++)
+			rte_memcpy(&ips[i], &ips[i + 1], DP_IPV6_ADDR_SIZE);
+}
+
+int dp_add_maglev_backend(struct lb_value *lbval, const uint8_t *new_server)
+{
+	int i, insert_pos = lbval->back_end_cnt;
+
+	if (lbval->back_end_cnt >= DP_LB_MAX_IPS_PER_VIP)
+		return DP_ERROR;
+
+	/* Find the correct position to insert the new server */
+	for (i = 0; i < lbval->back_end_cnt; i++) {
+		if (dp_is_ip_greater(lbval->back_end_ips[i], new_server)) {
+			insert_pos = i;
+			break;
+		}
+	}
+
+	dp_shift_back_end_ips(lbval->back_end_ips, insert_pos, lbval->back_end_cnt, DP_SHIFT_UP);
+	rte_memcpy(&lbval->back_end_ips[insert_pos], new_server, DP_IPV6_ADDR_SIZE);
+	lbval->back_end_cnt++;
+
+	return dp_maglev_calc_hash_lookup_table(lbval);
+}
+
+int dp_delete_maglev_backend(struct lb_value *lbval, const uint8_t *delete_server)
+{
+	int i;
+
+	for (i = 0; i < lbval->back_end_cnt; i++) {
+		if (memcmp(&lbval->back_end_ips[i], delete_server, DP_IPV6_ADDR_SIZE) == 0) {
+			dp_shift_back_end_ips(lbval->back_end_ips, i, lbval->back_end_cnt - 1, DP_SHIFT_DOWN);
+			memset(&lbval->back_end_ips[lbval->back_end_cnt - 1], 0, DP_IPV6_ADDR_SIZE);
+			lbval->back_end_cnt--;
+			break;
+		}
+	}
+
+	if (lbval->back_end_cnt == 0)
+		return DP_OK;
+	else
+		return dp_maglev_calc_hash_lookup_table(lbval);
+}
+
```

Comment:
Hopefully I'm right and this means no offense, but this code is surely based on a paper or some other research telling you how best to implement this algorithm. Which is perfectly fine, just that it is hard to understand due to lack of deeper understanding on the reader's part. 
I do not think we should pollute the code with comments, but a link or a reference of the implementation paper at the top would be a great addition.

Reply:
Put a comment at the beginning of the file that this code is based on maglev paper. Googling it should not be a problem, if someone wants to know more.